### PR TITLE
Electron 24 - protocol handler

### PR DESCRIPTION
**Pull request by VishwasShashidhar**
Implemented the protocol handler functionality on electron app for the following use cases:

1. open IM
2. open chat room
3. open user profile
4. use cases handling opening the electron symphony app in case it is already not open and loading the appropriate stream once the bridge is loaded.

### CODE REVIEW

**Comment by VishwasShashidhar** [comment]
@lneir @rbarros-daitan  Can you review this please?

**Comment by VishwasShashidhar** [comment]
I've pushed a new commit that adds support for windows. Apparently, electron doesn't emit 'open-url' event on windows. So, we've had to work around this by capturing the process argument listening on the callback of the make single instance function for when the app is already open. In case the app is not open, we listen on the 'on-ready' event and process the protocol action.

There is an open pull request on electron repo that emits the 'open-url' event even for windows, maybe once that is merged, we can use a single event to process the protocol action for both mac and windows.

More details here -> https://github.com/electron/electron/issues/9172 & https://github.com/electron/electron/pull/8052

File: js/main.js [hunk 1]
```diff
@@ -50,52 +58,61 @@ var symphonyAutoLauncher = new AutoLaunch({
 app.on('ready', getUrlAndOpenMainWindow);
 
 function getUrlAndOpenMainWindow() {
+
+    if (process.platform === 'win32') {
```
Comment (lneir) [inline js/main.js]: why only win32 here?  please add documentation.

File: js/main.js [hunk 2]
```diff
@@ -21,15 +26,18 @@ require('./memoryMonitor.js');
...
     let mainWin = windowMgr.getMainWindow();
     if (mainWin) {
+        isAppAlreadyOpen = true;
         if (mainWin.isMinimized()) {
             mainWin.restore();
         }
         mainWin.focus();
     }
+    // Handle protocol action if necessary
+    processProtocolAction(argv);
```
Comment (lneir) [inline js/main.js]: shouldn't this be moved to line 33?  also why this line and line 62 both?

File: js/main.js [hunk 3]
```diff
@@ -146,3 +193,15 @@ app.on('activate', function () {
         windowMgr.showMainWindow();
     }
 });
+
+app.setAsDefaultProtocolClient('symphony');
+
+app.on('open-url', function (event, url) {
+    if (!isAppAlreadyOpen) {
```
Comment (lneir) [inline js/main.js]: shouldn't this only be hit if app is already open?

File: js/mainApiMgr.js [hunk 4]
```diff
@@ -89,6 +95,11 @@ electron.ipcMain.on(apiName, (event, arg) => {
         log.setLogWindow(event.sender);
     }
 
+    if (arg.cmd === apiCmds.registerProtocolHandler) {
```
Comment (lneir) [inline js/mainApiMgr.js]: nit-picky but move this up to be after line 76 (close to other).

File: js/protocolHandler/index.js [hunk 5]
```diff
@@ -0,0 +1,57 @@
+/**
```
Comment (lneir) [inline js/protocolHandler/index.js]: don't need lines 1 to 3.

File: js/protocolHandler/index.js [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+ * Created by vishwas on 04/05/17.
+ */
+'use strict';
+
+let protocolWindow;
+var protocolUrl;
```
Comment (lneir) [inline js/protocolHandler/index.js]: why not use let?

File: js/main.js (hunk 1 above)
Comment (VishwasShashidhar) [inline js/main.js]: added.

File: js/main.js (hunk 2 above)
Comment (VishwasShashidhar) [inline js/main.js]: not exactly. `app.singleInstance` is called when we already have the app running, that means, we'll have the main window initialized in which the control will not go into the `if` condition. So, the `processProtocolAction` should be where it is now.

File: js/main.js (hunk 3 above)
Comment (VishwasShashidhar) [inline js/main.js]: that is what I had thought as well until I started testing and found out that electron emits this event before the 'ready' event when the app is not open. So, we'll need to handle both the cases here.

File: js/protocolHandler/index.js (hunk 5 above)
Comment (VishwasShashidhar) [inline js/protocolHandler/index.js]: ya, removed it. IDE auto-generated.

File: js/protocolHandler/index.js (hunk 6 above)
Comment (VishwasShashidhar) [inline js/protocolHandler/index.js]: makes sense. I've changed it now.

**Comment by VishwasShashidhar** [comment]
@lneir Please review now.

File: js/main.js [hunk 7]
```diff
@@ -146,3 +203,9 @@ app.on('activate', function () {
         windowMgr.showMainWindow();
     }
 });
+
+app.setAsDefaultProtocolClient('symphony');
+
+app.on('open-url', function (event, url) {
+    handleProtocolAction(url);
```
Comment (lneir) [inline js/main.js]: so on mac is this fired both when app is not running and running?

Comment (lneir) [inline js/main.js]: also noticed in documentation here: https://electron.atom.io/docs/api/app/#event-open-url-macos that something might be needed in plist???

File: js/main.js [hunk 8]
```diff
@@ -146,3 +203,9 @@ app.on('activate', function () {
         windowMgr.showMainWindow();
     }
 });
+
+app.setAsDefaultProtocolClient('symphony');
+
+app.on('open-url', function (event, url) {
```
Comment (lneir) [inline js/main.js]: as note that this event is only supported on mac.

File: js/main.js (hunk 7 above)
Comment (VishwasShashidhar) [inline js/main.js]: yes, it is triggered for both when the app is running and not running. The plist thing is taken care of automatically by the `app.setAsDefaultProtocolClient('symphony');` call.

File: js/main.js (hunk 8 above)
Comment (VishwasShashidhar) [inline js/main.js]: correct, only supported on mac currently.

**Pull request merged**